man/man1/velmap.1
.TH VELMAP 1NEMO "4 February 2011"
.SH NAME
velmap \- various galactic velocity field mapping routines
.SH SYNOPSIS
\fBvelmap\fP [parameter=value]
.SH DESCRIPTION
VELMAP has various mapping routines that can be done on
galactic velocity field.  The velocity field can be from
observations, or a model.
.PP
In practice this also means that on a set of projected
rings, defined by their radii, a center, position angle and inclination,
various computations can be done and tabulated. 
.PP
An output map can be generated with amusing quantities. 

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fP
input velocity field, in \fIimage(5NEMO)\fP format. 
.TP
\fBradii=\fP
radii of the ring boundaries. For N rings, N+1 values need to be give.
They need to be sorted by increasing radius. 
.TP
\fBpa=\fP
position angle of disk, in degrees. [0]
.TP
\fBinc=\fP
inclination angle of disk, in degrees [45]   
.TP
\fBvsys=\fP
systemic velocity [0]     
.TP
\fBcenter=\fP
rotation center. Default is the mapcenter ((Nx-1)/2,Ny-1)/2)
where 0,0=center of lower left pixel.
.TP
\fBden=\fP
input density image, only needed if surface brightness (d) given in 
an output mode. Default: not used.
\fBmode=\fP.
Default: none.
.TP
\fBfrang=\fP
free angle around minor axis (2*frang is the total angle around
the minor axis used).
[Default: 0]
.TP
\fBblank=\fP
Value of the blank pixel to be ignored. Default: 0.0
.TP
\fBmode=\fP
Output mode for the output data file. Valid are:
\fBv\fP, 
\fBv/r\fP (angular velocity),
\fBv*d/r\fP (angular velocity times surface brightness)
.TP
\fBout=\fP
Optional output file.  See \fBmode=\fP which quantity. Default: none
.TP
\fBtab=\fP
Optional output table file. This will contain the 
radius, rotation speed, dispersion in the ring and some
other values.

.SH "SEE ALSO"
rotcur(1NEMO), velfit(1NEMO), runvelfitss07(1NEMO), rotcurshape(1NEMO)
.SH FILES
src/image/rotcur : velmap.c
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +2.0i +4.0i
3-feb-2011	created for Nurur	PJT
.fi


src/image/rotcur/velmap.c
/*
 * VELMAP: 
 *
 *	VELMAP performs various mapping functions on a projected
 *      galactic velocity field.
 *      
 *     3-feb-2011   New task, cloned off velfit, for Nurur          PJT
 *    10-feb-2011   Cleaned up, aligned with velfit                 PJT
 *
 */

#include <nemo.h>
#include <image.h>

string defv[] = {
  "in=???\n       input velocity field",
  "radii=\n       radii of the ring boundaries (Nring+1)",
  "pa=0\n         position angle of disk",
  "inc=45\n       inclination angle of disk",
  "vsys=0\n       systemic velocity",
  "center=\n      rotation center (mapcenter if left blank, 0,0=lower left)",
  "den=\n         input density image, unity if left blank",
  "frang=0\n      free angle around minor axis (2*frang is the total)",
  "blank=0.0\n    Value of the blank pixel to be ignored",
  "mode=v\n       Output mode (v, v/r, d*v/r, ...)",
  "out=\n         Optional output map",
  "tab=\n         Optional output table",
  "VERSION=0.4\n  10-feb-2011 PJT",
  NULL,
};


string usage="various mapping functions on velocity fields";

string cvsid="$Id$";


#ifndef MAXRING
#define MAXRING    2048
#endif

bool Qden = FALSE;
bool Qout = FALSE;
bool Qtab = FALSE;
imageptr denptr = NULL, velptr = NULL, outptr = NULL;

real rad[MAXRING];
int nrad;

int  pixe[MAXRING];
real vsum[MAXRING], vsqu[MAXRING], wsum[MAXRING];
real vrot[MAXRING];

real pa, inc, vsys, xpos, ypos;
real undf;

               /*  0 1   2    */
string outmodes = "v,v/r,d*v/r";

int string_index(string options, string s)
{
  int i=0;
  string *sa = burststring(options,",");

  while (sa[i]) {
    if (streq(sa[i],s))  {
      freestrings(sa);
      return i;
    }
    i++;
  }
  freestrings(sa);	   
  return -1;
}

int ring_index(int n, real *r, real rad)
{
  int i;
  if (rad < r[0]) return -1;
  if (rad > r[n-1]) return -2;
  for (i=0;i<n;i++)
    if (rad >= r[i] && rad < r[i+1]) return i;
  error("ring_index: should never gotten here %g in [%g : %g]",
	rad,r[0],r[n-1]);
}

nemo_main()
{
  stream denstr, velstr, outstr, tabstr;
  real center[2], cospa, sinpa, cosi, sini, cost, costmin, x, y, xt, yt, r, den;
  real vr, wt, frang, dx, dy, xmin, ymin, rmin, rmax, ave, tmp, rms;
  real sincosi, cos2i, tga, dmin, dmax, dval, dr, area, fsum1, fsum2;
  int i, j, k, nx, ny, ir, nring, nundf, nout, nang, nsum;
  string outmode;
  int mode = -1;

  velstr = stropen(getparam("in"),"r");

  read_image(velstr,&velptr);
  nx = Nx(velptr);
  ny = Ny(velptr);

  if (hasvalue("out")) {
    outmode = getparam("mode");
    mode = string_index(outmodes, outmode);
    if (mode < 0) error("Illegal mode=%s [%d], valid:",outmode,mode,outmodes);
    warning("New out= mode mode=%s [%d]",outmode,mode);
    Qout = TRUE;
    outstr = stropen(getparam("out"),"w");
    copy_image(velptr,&outptr);
  } 

  if (hasvalue("den")) {
    Qden = TRUE;
    denstr = stropen(getparam("den"),"r");
    read_image(denstr,&denptr);
  } else if (mode==2)
    error("Need den=");


  if (hasvalue("tab")) {
    Qtab = TRUE;
    tabstr = stropen(getparam("tab"),"w");
  } 

  nrad = nemoinpd(getparam("radii"),rad,MAXRING);
  if (nrad < 2) error("got no rings (%d), use radii=",nrad);
  nring = nrad-1;
  if (hasvalue("center")) {
    if (nemoinpd(getparam("center"),center,2) != 2)
      error("not enuf values for center=, need 2");
    xpos = center[0];
    ypos = center[1];
  } else {
    xpos = (Nx(velptr)-1.0)/2.0;
    ypos = (Ny(velptr)-1.0)/2.0;
  }
  pa    = getdparam("pa");
  inc   = getdparam("inc");
  vsys  = getdparam("vsys");
  undf  = getdparam("blank");
  frang = getdparam("frang");

  cospa   = cos(pa*PI/180.0);
  sinpa   = sin(pa*PI/180.0);
  sini    = sin(inc*PI/180.0);
  cosi    = cos(inc*PI/180.0);
  costmin = sin(frang*PI/180.0);
  sincosi = sini*cosi;
  cos2i   = cosi*cosi;
    
  for (i=0; i<nring; i++)
    pixe[i] = vsum[i] = vsqu[i] = wsum[i] = 0;
  nundf = nout = nang = 0;

  ymin = Ymin(velptr);
  xmin = Xmin(velptr);
  dx = Dx(velptr);       dx = ABS(dx);    dx = -dx;
  dy = Dy(velptr);       dy = ABS(dy);
  rmin = -nx*dx*10.0;
  rmax = 0.0;
  dprintf(0,"Map %d x %d pixels, size %g x %g\n",
	  Nx(velptr), Ny(velptr), -dx*Nx(velptr), dy*Ny(velptr));
  dprintf(0,"Pixel size: %g x %g\n",-dx, dy);

  dmin = dmax = vsys;

  /* loop over the map, accumulating data for fitting process */

  for (j=0; j<ny; j++) {
    y = (j-ypos)*dy;
    for (i=0; i<nx; i++) {
      if (MapValue(velptr,i,j) == undf) {
	nundf++;
	if (Qout) MapValue(outptr,i,j) = undf;
	continue;
      }
      x = (i-xpos)*dx;
      yt = x*sinpa + y*cospa;      /* major axis now along Y  */
      xt = x*cospa - y*sinpa;      /* minor axis along X      */
      xt /= cosi;                  /* deproject to the circle */
      r  = sqrt(xt*xt+yt*yt);      /* radius in the disk      */
      rmin = MIN(r,rmin);
      rmax = MAX(r,rmax);
      ir = ring_index(nrad,rad,r);
      dprintf(2,"r=%g ir=%d  (x,y)=%g,%g  (xt,yt)=%g,%g\n",
	      r,ir,x,y,xt,yt);
      if (ir < 0) {
	nout++;
	continue;
      }
      cost = yt/r;
      dval = MapValue(velptr,i,j);
      
      if (mode==1)
	dval /= r;
      else if (mode==2)
	dval *= MapValue(denptr,i,j) / r;

      if (outptr) {
	if  (ABS(cost) > costmin) {
	  MapValue(outptr,i,j) = dval;
	  if (dval > dmax) dmax = dval;
	  if (dval < dmin) dmin = dval;
	} else {
	  MapValue(outptr,i,j) = undf;
	}
      }

      /* now some ring accumulation, remnant of the velfit fitting */

      if (ABS(cost) > costmin) {
	pixe[ir] += 1;
	wsum[ir] += 1.0;
	vsum[ir] += dval;
	vsqu[ir] += dval*dval;
      } else
	nang++;
    } /* i */
  } /* j */

  /* write output map(s), if needed */
  if (Qout) {
    dprintf(0,"Data min/max = %g %g\n",dmin,dmax);    
    MapMin(outptr) = dmin;
    MapMax(outptr) = dmax;
    write_image(outstr,outptr);
  }

  /* report on the rings */

  if (Qtab) {
    fprintf(tabstr,"# r I rms I_sum1 I_sum2 i_ring Npoints\n");
    fsum1 = 0.0;  /* this will count rings with average values */
    fsum2 = 0.0;  /* this will count up flux whenever it fell in a ring */
    nsum = 0;
    for (i=0; i<nring; i++) {
      if (wsum[i] == 0.0) continue;
      nsum++;
      r = 0.5*(rad[i] + rad[i+1]);
      dr = rad[i+1] - rad[i];
      area = PI*(sqr(rad[i+1]) - sqr(rad[i]));
      ave = vsum[i]/wsum[i];
      rms = vsqu[i]/wsum[i]-ave*ave;
      if (rms <  0) rms=0.0;
      rms = sqrt(rms);
      fsum1 += ave*area;
      fsum2 += vsum[i];
      
      fprintf(tabstr,"%g %g %g  %g %g  %d %d\n",
	      r,ave,rms,fsum1,fsum2,i+1,pixe[i]);
    }
  }
  dprintf(0,"Nundf=%d/%d Nout=%d Nang=%d (sum=%d)\n",
	  nundf,nx*ny,nout,nang,nout+nundf+nang);
  dprintf(0,"Rmin/max = %g %g\n",rmin,rmax);

}

